jakttest/jakttest.jakt

```diff
@@ -334,31 +405,124 @@ struct Options {
     }
 }
 
-enum TestStage {
-    TranspileJakt
-    CompileCpp
-    TestRun
+class PipeOutput {
+    buffer: [u8]
+    wrote_count: usize
+    fd: i32
+
+    public function init(fd: i32, buffer: [u8]) throws => 
+        PipeOutput(
+            buffer
+            wrote_count: 0
+            fd)
+
+    function finished_writing(this) => .wrote_count == .buffer.size()
+
+    public function execute_write(mut this) throws {
+        if .finished_writing() {
+            return
+        }
+
+        // NOTE: mut because codegen prepends const to the type, so pointer
+        // variables of the type `raw T` declared with `let` end up being const T *
+        // as their type.
+        mut write_ptr = get_array_unsafe_pointer(.buffer)
+        let written_count = os::write_nonblocking(fd: .fd, buffer: write_ptr, max_len: .buffer.size() - .wrote_count)
+
+        if written_count.has_value() {
+            .wrote_count += written_count!
+            // since we've finished writing, we'll close the file
+            // descriptor so that the reading end can get a "read 0"
+            if .finished_writing() {
+                os::close(fd: .fd)
+            }
+        }
+
+    }
+
+    public function release(this) throws -> [u8] {
+        if not .finished_writing() {
+            os::close(fd: .fd)
+        }
+        return .buffer
+    }
+}
 
-    function equals(this, anon other: TestStage) => match this {
-        TranspileJakt => other is TranspileJakt
-        CompileCpp => other is CompileCpp
-        TestRun => other is TestRun
+// raw buffer size: usize = 4096 (4KiB)
+
+class CollectedReader {
+    collected_buffer: [u8]
+    raw_buffer: raw u8
+    raw_buffer_read: u16
+    fd: i32
+
+    public function init(fd: i32) throws -> CollectedReader {
+        // NOTE: mut so codegen does not declare it as const u8 *
+        mut raw_buffer = libc::malloc(size: 4096)
```

Comment:
this feels very awkward. does Jakt runtime not have AK::ByteBuffer?

Reply:
Probably it's there but I just didn't know of it. Anyway, this is nowhere near finished. The scheduler gets stuck after the first chunk of jobs finishes, panics sometimes (maybe it's the queue implementation, I don't know), fails to properly collect stdout/stderr over the pipes (thus marking tests as failed because it couldn't find the error). Plus, the fork()/dup2()/exec() has to be replaced with an API that implements the needed `posix_spawn`* functions while allowing a similar approach for Windows, since the Windows approach goes through a 'builder pattern' in a similar way that `posix_spawn` functions are meant to be used.

Reply:
I'm going to throw out the pipes and just create the `.err`/`.out`/`.cpp` files directly and redirect the output to them, saving just their handle. This will just need a little add-on to `File` to get `fileno` functionality, and will remove the polling of buffers that is not even working properly.